File: lingting-spring-redis/src/main/java/live/lingting/spring/redis/script/RepeatRedisScript.java
package live.lingting.spring.redis.script;

import live.lingting.spring.redis.Redis;
import lombok.Getter;
import org.redisson.api.RScript;
import org.redisson.api.RedissonClient;
import org.springframework.data.redis.connection.RedisConnection;
import org.springframework.data.redis.connection.ReturnType;
import org.springframework.data.redis.core.script.DigestUtils;
import org.springframework.util.Assert;
import org.springframework.util.StringUtils;

import java.util.List;

/**
 * @author lingting 2024-04-17 16:19
 */
@Getter
public class RepeatRedisScript<T> {

	private final String source;

	private final String sha1;

	private final ReturnType type;

	private boolean load = false;

	public RepeatRedisScript(String source) {
		this(source, ReturnType.STATUS);
	}

	public RepeatRedisScript(String source, Class<T> resultType) {
		this(source, ReturnType.fromJavaType(resultType));
	}

	public RepeatRedisScript(String source, ReturnType type) {
		Assert.state(StringUtils.hasText(source), "Redis script source must not be empty");
		this.source = source;
		this.sha1 = DigestUtils.sha1DigestAsHex(source);
		this.type = type;
	}

	public T execute(List<String> keys, Object... args) {
		return Redis.instance().scriptExecutor().execute(this, keys, args);
	}

	public T execute(RedisConnection connection, List<String> keys, Object... args) {
		return Redis.instance().scriptExecutor().execute(connection, this, keys, args);
	}

	public void load(RedissonClient client) {
		RScript script = client.getScript();
		script.scriptLoad(source);
		this.load = true;
	}

}

File: lingting-spring-redis/src/main/kotlin/live/lingting/spring/redis/script/RepeatRedisScript.kt
package live.lingting.spring.redis.script

import java.util.concurrent.ConcurrentHashMap
import live.lingting.spring.redis.Redis
import org.springframework.data.redis.connection.ReturnType
import org.springframework.data.redis.core.script.DigestUtils
import org.springframework.data.redis.core.script.RedisScript

/**
 * @author lingting 2024-04-17 16:19
 */
class RepeatRedisScript<T> @JvmOverloads constructor(
    source: String,
    val type: ReturnType = ReturnType.STATUS
) {

    companion object {

        @JvmField
        val CACHE = ConcurrentHashMap<RedisScript<*>, RepeatRedisScript<*>>()

        @JvmStatic
        fun <T> of(source: RedisScript<T>): RepeatRedisScript<T> {
            return CACHE.computeIfAbsent(source) { RepeatRedisScript(source.scriptAsString, source.resultType) } as RepeatRedisScript<T>
        }

        fun <T> RedisScript<T>.repeat(): RepeatRedisScript<T> = of(this)

    }

    val source: String

    val sha1: String

    var load: Boolean = false

    var bytes: ByteArray? = null

    constructor(source: String, resultType: Class<T>?) : this(source, ReturnType.fromJavaType(resultType))

    init {
        require(source.isNotBlank()) { "Redis script source must not be empty" }
        this.source = source
        this.sha1 = DigestUtils.sha1DigestAsHex(source)
    }

    @JvmOverloads
    fun execute(keys: List<String> = emptyList(), vararg args: Any): T? {
        return execute(keys, args.toList())
    }

    @JvmOverloads
    fun execute(keys: List<String> = emptyList(), args: Collection<Any>): T? {
        return Redis.instance().script(this).execute(keys, args)
    }


}
